Allow types to determine how to decode themselves

mthjs opened the pull request:
When the destination of a decode implements the custom decoder interface, use said interface. Custom decoders take precedence over the default decoding rules.

This should resolve https://github.com/mitchellh/mapstructure/issues/115

Renamed the master -> main which closed [the pull-request](https://github.com/mitchellh/mapstructure/pull/287), sorry :sweat_smile:

Commit:
Allow types to determine how to decode themselves

wrouesnel commented:
I'd like to see this merged, though to offer one comment, the interface function - Decode() might be a bit too generic within the go-ecosystem? `MapStructureDecode` maybe?

andig commented:
+1 for allowing objects to decode themselves.

Commit:
Make the custom decoder more specific to mapstructure

Similar to the `UnarshalX` family of functions seen throughout
`encoding` and `encoding/json`, include the name of what is being
decoded into the function.

mthjs commented:
@wrouesnel I've applied your suggestion although I did swap the order. Like any of the standard-libraries `UnmarshalX` functions, I've opted to suffix the expected `Decode` function with `Mapstructure` instead of prefixing it.

stephen-slm commented:
@mitchellh  any interest in this being merged?

dplepage-dd commented:
Would it be possible to make this work in both directions, so that e.g. decoding a `Someone` into an `interface{}` produces a string (the same way `json` allows customizing both `MarshallJSON` and `UnmarshallJSON`)?

There would be some ambiguity, I guess, since you could conceivably call `Decode(x,y)` with `x` and `y` both being types with custom encoder/decoders, but in practice I think it's rarely used as a generic converter, it's almost always either decoding from a map structure to a custom type or encoding from a custom type back to a map structure, so it would probably be fine to just error out in case of conflicting custom handlers.

dplepage-dd commented on mapstructure.go:
```diff
@@ -453,6 +453,13 @@ func (d *Decoder) decode(name string, input interface{}, outVal reflect.Value) e
+	if outVal.CanAddr() {
+		v := outVal.Addr()
+		if u, ok := v.Interface().(decoder); ok {
+			return u.DecodeMapstructure(input)
```
I think it might make more sense to put this after the `DecodeHook`, so that any custom preprocessing can still happen.

pcfreak30 commented:
Bumping this b/c this would probably simplify my code a lot to use stdlib marshal interfaces and only 1 hook for it all.

buzzdan commented:
@pcfreak30 @mitchellh any updates on this one ?

pcfreak30 commented:
> @pcfreak30 @mitchellh any updates on this one ?

Not up to me. I have forked to continue my development, there is enough effort as-is for @mitchellh to make a PR and add this if he decides to.
